gnomad_qc/v4/analyses/grpmax_comps.py

```diff
@@ -0,0 +1,175 @@
+        )
+    ht = ht.annotate(
+        eur_AF=hl.sum(
+            eur_indices.map(
+                lambda f: hl.if_else(hl.is_defined(ht.freq[f].AC), ht.freq[f].AC, 0)
+            )
+        )
+        / hl.sum(
+            eur_indices.map(
+                lambda f: hl.if_else(hl.is_defined(ht.freq[f].AN), ht.freq[f].AN, 0)
+            )
+        )
+    )
+    ht = ht.filter(hl.is_defined(ht.eur_AF))
+
+    return ht
+
+
+def filter_to_threshold(
+    ht: hl.Table,
+    af_threshold: float = 0.001,
+    version: str = "v4",
+):
+    """
+    Filter to variants where eur AF is < threshold while grpmax > threshold.
+
+    :param ht: gnomAD release Table
+    :param af_threshold: AF dividing threshold
+    :param version: gnomAD version
+    :return: Table filtered to variants meething threshold specifications
+    """
+    if version == "v4":
+        grpmax_expr = ht.grpmax.gnomad
+    else:
+        grpmax_expr = ht.popmax[0]
+    ht = ht.filter((grpmax_expr.AF > af_threshold) & (ht.eur_AF < af_threshold))
+    return ht
+
+
+def version_stats(
+    ht: hl.Table, version: str = "v4"
+) -> Dict[str, Dict[str, Dict[str, int]]]:
+    """
+    Calculate grpmax stats for a given gnomAD version.
+
+    :param ht: gnomAD release Table
+    :param version: gnomAD version
+    :return: Dictionary of grpmax stats
+    """
+    logger.info(f"Calculating grpmax stats for {version}")
+    # Group results into nfe and all eur grps
+    results_by_eur_grping = {}
+    # Filter to only non-synonymous terms CSQ_CODING_HIGH_IMPACT +
+    # CSQ_CODING_MEDIUM_IMPACT
+    ht = ht.filter(NS_CONSEQ_TERMS.contains(ht.vep.most_severe_consequence))
+    # Filter to PASS variants only
+    ht = ht.filter(ht.filters.length() == 0)
+    # Iterate through just nfe group and all eur grp for calculating eur AF
+    for grp_id, grps in EUR_GRPS.items():
+        # Get european frequency by calculating the cumulative AF in the passed
+        # europe genetic ancestry list
+        p_ht = get_eur_freq(ht, eur_grps=grps, version=version)
+
+        grpmax_by_thresholds = {}
+        for threshold in AF_THRESHOLDS:
+            # Filter to keep only variants that split at the AF threshold
+            t_ht = filter_to_threshold(p_ht, threshold, version=version)
+            if version == "v4":
+                grpmax_ga_expr = t_ht.grpmax.gnomad.gen_anc
+            else:
+                grpmax_ga_expr = t_ht.popmax[0].pop
+
+            t_ht = t_ht.filter(DIVERSE_GRPS.contains(grpmax_ga_expr))
+            if version == "v4":
+                grpmax_ga_expr = t_ht.grpmax.gnomad.gen_anc
+            else:
+                grpmax_ga_expr = t_ht.popmax[0].pop
+            # For each diverse genetic ancestry group, aggregate the number of
+            # variants where that group is grpmax
+            grpmax_by_thresholds[threshold] = t_ht.aggregate(
+                hl.agg.counter(grpmax_ga_expr)
+            )
+        results_by_eur_grping[grp_id] = grpmax_by_thresholds
```

Comment:
minor comment that adding the total number of variants to this dict would be helpful